Consider DBNull.Value and NullString.Value the same as $null

vexx32 opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

Closes #9561 

- Adds `LanguagePrimitives.IsNullLike()` method to account for `DBNull.Value` and `NullString.Value` so that they can be considered the same as a null value where sensible in PowerShell.
- Updates `-ne` and `-eq` binders to treat `DBNull.Value` and `NullString.Value` as equal to null/AutomationNull.
- Update code paths for comparing objects in LanguagePrimitives to ensure consistency with how the `-eq` and `-ne` binders work when calling LanguagePrimitives methods to do the comparisons.
- Make `LanguagePrimitives.IsNull()` and `LanguagePrimitives.IsNullLike()` public methods.
- Added tests for null behaviours in `NullRepresentatives.Tests.ps1`

## PR Context

DBNull.Value and NullString.Value are generally considered to be equivalent to null, and it makes the most sense to have PowerShell recognise them as such instead of treating them as strange class objects. See linked issue for more details.

/cc @daxian-dbw 

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

PoshChan commented:
@vexx32, your last commit had 6 failures in `PowerShell-CI-macos`
(These are 5 of the failures)

Alias tests.Export-Alias literal path.for CSV
```powershell
Expected <empty>, but got $null.
at <ScriptBlock>, /Users/vsts/agent/2.150.3/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/alias.tests.ps1: line 57
57:                     $exportAliasError.FullyqualifiedErrorId | Should -Be $test.expectedError
```
Alias tests.Export-Alias literal path.for PS1
```powershell
Expected <empty>, but got $null.
at <ScriptBlock>, /Users/vsts/agent/2.150.3/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/alias.tests.ps1: line 57
57:                     $exportAliasError.FullyqualifiedErrorId | Should -Be $test.expectedError
```
Alias tests.Export-Alias literal path.for Null
```powershell
Expected strings to be the same, but they were different.
Expected length: 95
Actual length:   102
Strings differ at index 32.
Expected: '...ErrorNullN...'
But was:  '...ErrorEmpty...'
at <ScriptBlock>, /Users/vsts/agent/2.150.3/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/alias.tests.ps1: line 57
57:                     $exportAliasError.FullyqualifiedErrorId | Should -Be $test.expectedError
```
Alias tests.Import-Alias literal path.for Null
```powershell
Expected an exception, with FullyQualifiedErrorId 'ParameterArgumentValidationErrorNullNotAllowed,Microsoft.PowerShell.Commands.ImportAliasCommand' to be thrown, but the FullyQualifiedErrorId was 'ParameterArgumentValidationErrorEmptyStringNotAllowed,Microsoft.PowerShell.Commands.ImportAliasCommand'. from /Users/vsts/agent/2.150.3/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/alias.tests.ps1:110 char:45
    +                 { Import-Alias -LiteralPath $test.testFile -ErrorActi ...
    +                                             ~~~~~~~~~~~~~~
at <ScriptBlock>, /Users/vsts/agent/2.150.3/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/alias.tests.ps1: line 110
110:                 { Import-Alias -LiteralPath $test.testFile -ErrorAction SilentlyContinue } | Should -Throw -ErrorId $test.expectedError
```
CliXml test.Export-CliXml.with path as Null
```powershell
Expected an exception, with FullyQualifiedErrorId 'ParameterArgumentValidationErrorNullNotAllowed,Microsoft.PowerShell.Commands.ExportClixmlCommand' to be thrown, but the FullyQualifiedErrorId was 'ParameterArgumentValidationErrorEmptyStringNotAllowed,Microsoft.PowerShell.Commands.ExportClixmlCommand'. from /Users/vsts/agent/2.150.3/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1:60 char:55
    + ...          { Export-Clixml -Depth 1 -LiteralPath $test.testFile -InputO ...
    +                                                    ~~~~~~~~~~~~~~
at <ScriptBlock>, /Users/vsts/agent/2.150.3/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1: line 60
60:                 { Export-Clixml -Depth 1 -LiteralPath $test.testFile -InputObject $test.inputObject -Force } | Should -Throw -ErrorId $test.expectedError
```

PoshChan commented:
@vexx32, your last commit had 6 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

Alias tests.Export-Alias literal path.for CSV
```powershell
Expected <empty>, but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\alias.tests.ps1: line 57
57:                     $exportAliasError.FullyqualifiedErrorId | Should -Be $test.expectedError
```
Alias tests.Export-Alias literal path.for PS1
```powershell
Expected <empty>, but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\alias.tests.ps1: line 57
57:                     $exportAliasError.FullyqualifiedErrorId | Should -Be $test.expectedError
```
Alias tests.Export-Alias literal path.for Null
```powershell
Expected strings to be the same, but they were different.
Expected length: 95
Actual length:   102
Strings differ at index 32.
Expected: '...ErrorNullN...'
But was:  '...ErrorEmpty...'
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\alias.tests.ps1: line 57
57:                     $exportAliasError.FullyqualifiedErrorId | Should -Be $test.expectedError
```
Alias tests.Import-Alias literal path.for Null
```powershell
Expected an exception, with FullyQualifiedErrorId 'ParameterArgumentValidationErrorNullNotAllowed,Microsoft.PowerShell.Commands.ImportAliasCommand' to be thrown, but the FullyQualifiedErrorId was 'ParameterArgumentValidationErrorEmptyStringNotAllowed,Microsoft.PowerShell.Commands.ImportAliasCommand'. from D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\alias.tests.ps1:110 char:45
    +                 { Import-Alias -LiteralPath $test.testFile -ErrorActi ...
    +                                             ~~~~~~~~~~~~~~
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\alias.tests.ps1: line 110
110:                 { Import-Alias -LiteralPath $test.testFile -ErrorAction SilentlyContinue } | Should -Throw -ErrorId $test.expectedError
```
CliXml test.Export-CliXml.with path as Null
```powershell
Expected an exception, with FullyQualifiedErrorId 'ParameterArgumentValidationErrorNullNotAllowed,Microsoft.PowerShell.Commands.ExportClixmlCommand' to be thrown, but the FullyQualifiedErrorId was 'ParameterArgumentValidationErrorEmptyStringNotAllowed,Microsoft.PowerShell.Commands.ExportClixmlCommand'. from D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\clixml.tests.ps1:60 char:55
    + ...          { Export-Clixml -Depth 1 -LiteralPath $test.testFile -InputO ...
    +                                                    ~~~~~~~~~~~~~~
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\clixml.tests.ps1: line 60
60:                 { Export-Clixml -Depth 1 -LiteralPath $test.testFile -InputObject $test.inputObject -Force } | Should -Throw -ErrorId $test.expectedError
```

PoshChan commented:
@vexx32, your last commit had 6 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

Alias tests.Export-Alias literal path.for CSV
```powershell
Expected <empty>, but got $null.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/alias.tests.ps1: line 57
57:                     $exportAliasError.FullyqualifiedErrorId | Should -Be $test.expectedError
```
Alias tests.Export-Alias literal path.for PS1
```powershell
Expected <empty>, but got $null.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/alias.tests.ps1: line 57
57:                     $exportAliasError.FullyqualifiedErrorId | Should -Be $test.expectedError
```
Alias tests.Export-Alias literal path.for Null
```powershell
Expected strings to be the same, but they were different.
Expected length: 95
Actual length:   102
Strings differ at index 32.
Expected: '...ErrorNullN...'
But was:  '...ErrorEmpty...'
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/alias.tests.ps1: line 57
57:                     $exportAliasError.FullyqualifiedErrorId | Should -Be $test.expectedError
```
Alias tests.Import-Alias literal path.for Null
```powershell
Expected an exception, with FullyQualifiedErrorId 'ParameterArgumentValidationErrorNullNotAllowed,Microsoft.PowerShell.Commands.ImportAliasCommand' to be thrown, but the FullyQualifiedErrorId was 'ParameterArgumentValidationErrorEmptyStringNotAllowed,Microsoft.PowerShell.Commands.ImportAliasCommand'. from /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/alias.tests.ps1:110 char:45
    +                 { Import-Alias -LiteralPath $test.testFile -ErrorActi ...
    +                                             ~~~~~~~~~~~~~~
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/alias.tests.ps1: line 110
110:                 { Import-Alias -LiteralPath $test.testFile -ErrorAction SilentlyContinue } | Should -Throw -ErrorId $test.expectedError
```
CliXml test.Export-CliXml.with path as Null
```powershell
Expected an exception, with FullyQualifiedErrorId 'ParameterArgumentValidationErrorNullNotAllowed,Microsoft.PowerShell.Commands.ExportClixmlCommand' to be thrown, but the FullyQualifiedErrorId was 'ParameterArgumentValidationErrorEmptyStringNotAllowed,Microsoft.PowerShell.Commands.ExportClixmlCommand'. from /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1:60 char:55
    + ...          { Export-Clixml -Depth 1 -LiteralPath $test.testFile -InputO ...
    +                                                    ~~~~~~~~~~~~~~
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1: line 60
60:                 { Export-Clixml -Depth 1 -LiteralPath $test.testFile -InputObject $test.inputObject -Force } | Should -Throw -ErrorId $test.expectedError
```

vexx32 commented:
@daxian-dbw Given how PowerShell handles conversions from null -> string (yielding `string.Empty` if I'm not mistaken) it may not make sense to also handle `NullString.Value` as a true `null` here as well -- either that or I need to be more careful with the handling in the string conversion paths so that a null string is a valid result when given a "null" value.

See the failing tests, where it expects empty string because the class definition has a `[string]`-type property and it is converting "expected" values into empty strings as a result here.

What would be appropriate here, do you think?

daxian-dbw commented:
I think `[DBNull]::Value` and `[NullString]::Value` should be treated as null only when comparing with `$null`, but not in general.
With your change, `gm -InputObject ([System.DBNull]::Value)` will start to fail.

Quote @mklement0's comment in the original issue:
> I wouldn't frame the issue in terms of $null, but in terms of (possibly implicit) to-Boolean conversion.
> Secondarily, the question is how direct comparison with $null should be handled.

As for `LanguagePrimitive.IsNull`, I think it probably should be renamed to `IsLiterallyNull()` and keep the existing implementation unchanged.

vexx32 commented:
🤔 I guess that does make sense. I can do that.

Any chance you can point me in the general direction of where the direct comparison with $null would be implemented instead? 

Also, given that, would we consider `[DBNull]::Value -eq `[NullString]::Value`?

daxian-dbw commented:
> Also, given that, would we consider `[DBNull]::Value -eq [NullString]::Value`?

I don't think we should consider them equal. The following shows how they should behave in my opinion:
```
## Equality comparison with $null should return true
$null -eq [AutomationNull].Value
> $true

[AutomationNull].Value -eq $null
> $true

$null -eq [DBNull].Value
> $true

[DBNull].Value -eq $null
> $true

$null -eq [NullString]::Value
> $true

[NullString]::Value -eq $null
> $true

## Equality comparison between '[AutomationNull].Value', '[DBNull].Value', and '[NullString].Value' should return false.
[AutomationNull].Value -eq [DBNull].Value
> $false

[DBNull].Value -eq [NullString]::Value
> $false

[NullString]::Value -eq [AutomationNull].Value
> $false
```
/cc @mklement0 for comments :)

> would it make sense to make IsLiterallyNull() public 

I agree to make it public, but we may need to chew on the name a bit more :)

> Any chance you can point me in the general direction of where the direct comparison with $null would be implemented instead?

Sure, I will take a look at the code later today.

vexx32 commented:
Appreciate it! Personally I'm a fan of `IsActuallyNull()` 😉  (or maybe `IsTrueNull()`)

vexx32 committed:
:wrench: Update binders for comparisons to $null
Add exceptions to logic where DBNull and NullString are concerned.
This allows PS to consider both values as equivalent to $null when doing
comparisons with $null on both RHS and LHS.

mklement0 commented:
My vote is for `IsTrueNull()`.

@daxian-dbw, I agree with the proposed behavior, thanks for hashing it out.

For someone who needs to distinguish between `$null` and one of the null-like values in an expression context (as an alternative to using `Get-Member`), they can use:

```powershell
[NullString]::Value -is [NullString] 
[System.DBNull]::Value -is [System.DBNull] 
```

Unfortunately, we do _not_ get the same behavior with `[System.Management.Automation.Internal.AutomationNull]::Value`, because it is seemingly converted to true `$null` during parameter binding and in expressions:

```powershell
# FAILS - same as passing $null
Get-Member -InputObject ([System.Management.Automation.Internal.AutomationNull]::Value)

# $FALSE, paradoxically
[System.Management.Automation.Internal.AutomationNull]::Value -is [System.Management.Automation.Internal.AutomationNull]
```

Distinguishing between true `$null` and `[System.Management.Automation.Internal.AutomationNull]::Value` is currently only possible with obscure workarounds; e.g.:

```powershell
# Determine if a given variable $v is true $null or AutomationNull
$isAutomationNull = $null -eq $v -and @($v).Count -eq 0
$isTrueNull = $null -eq $v -and @($v).Count -eq 1
```

Related: https://github.com/PowerShell/PowerShell/issues/7049#issuecomment-399808799

vexx32 commented:
@daxian-dbw any luck finding an appropriate codepoint to check for this? I've made a couple attempts myself but I've not found anything that gets the right behaviour as yet.

daxian-dbw commented:
@vexx32 Sorry for being tardy. I have been focusing on some PSReadLine issues and lost track of this PR.
I think there are a few places that need to be updated for the desired behavior:

1. For binary comparison in language
https://github.com/PowerShell/PowerShell/blob/65a3e467cc95059d16f58174335e6dd3a58c459b/src/System.Management.Automation/engine/runtime/Binding/Binders.cs#L3024-L3041
AND
https://github.com/PowerShell/PowerShell/blob/65a3e467cc95059d16f58174335e6dd3a58c459b/src/System.Management.Automation/engine/runtime/Binding/Binders.cs#L3047-L3063

2. For static methods in `LanguagePrimitives`
```
LanguagePrimitives.Equals
LanguagePrimitives.IsTrue
LanguagePrimitives.Compare // and other compare/trycompare overloads, to handle comparing $null and the special NULL representatives.
Review all methods in `LangaugePrimitives.cs` that are related to null comparison in names, like 'CompareObjectToNull'
```

vexx32 committed:
:wrench: Update comparison logic for various nulls
Adds NullString.Value and DBNull.Value to comparison logic paths.

vexx32 committed:
:white_check_mark: Add tests for spec behaviour

PoshChan commented:
@vexx32, your last commit had 5 failures in `PowerShell-CI-windows`
Null Representatives.Comparisons with $null.{ [DBNull]::Value } should be equivalent to $null (RHS: $null)
```powershell
Expected $true, but got $false.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Scripting\NullRepresentatives.Tests.ps1: line 20
20:             $Value.InvokeReturnAsIs() -eq $null | Should -BeTrue
```
Null Representatives.Comparisons with $null.{ [NullString]::Value } should be equivalent to $null (RHS: $null)
```powershell
Expected $true, but got $false.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Scripting\NullRepresentatives.Tests.ps1: line 20
20:             $Value.InvokeReturnAsIs() -eq $null | Should -BeTrue
```
Null Representatives.Comparisons with other null representatives.{ [AutomationNull]::Value } should not be equal to { [DBNull]::Value }
```powershell
Expected $false, but got $true.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Scripting\NullRepresentatives.Tests.ps1: line 42
42:             $LHS.InvokeReturnAsIs() -eq $RHS.InvokeReturnAsIs() | Should -BeFalse
```
Null Representatives.Comparisons with other null representatives.{ [AutomationNull]::Value } should not be equal to { [NullString]::Value }
```powershell
Expected $false, but got $true.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Scripting\NullRepresentatives.Tests.ps1: line 42
42:             $LHS.InvokeReturnAsIs() -eq $RHS.InvokeReturnAsIs() | Should -BeFalse
```
Null Representatives.Comparisons with other null representatives.{ [DBNull]::Value } should not be equal to { [NullString]::Value }
```powershell
Expected $false, but got $true.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Scripting\NullRepresentatives.Tests.ps1: line 43
43:             $RHS.InvokeReturnAsIs() -eq $RHS.InvokeReturnAsIs() | Should -BeFalse
```

PoshChan commented:
@vexx32, your last commit had 5 failures in `PowerShell-CI-macos`
Null Representatives.Comparisons with $null.{ [DBNull]::Value } should be equivalent to $null (RHS: $null)
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/vsts/agent/2.152.1/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 20
20:             $Value.InvokeReturnAsIs() -eq $null | Should -BeTrue
```
Null Representatives.Comparisons with $null.{ [NullString]::Value } should be equivalent to $null (RHS: $null)
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/vsts/agent/2.152.1/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 20
20:             $Value.InvokeReturnAsIs() -eq $null | Should -BeTrue
```
Null Representatives.Comparisons with other null representatives.{ [AutomationNull]::Value } should not be equal to { [DBNull]::Value }
```powershell
Expected $false, but got $true.
at <ScriptBlock>, /Users/vsts/agent/2.152.1/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 42
42:             $LHS.InvokeReturnAsIs() -eq $RHS.InvokeReturnAsIs() | Should -BeFalse
```
Null Representatives.Comparisons with other null representatives.{ [AutomationNull]::Value } should not be equal to { [NullString]::Value }
```powershell
Expected $false, but got $true.
at <ScriptBlock>, /Users/vsts/agent/2.152.1/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 42
42:             $LHS.InvokeReturnAsIs() -eq $RHS.InvokeReturnAsIs() | Should -BeFalse
```
Null Representatives.Comparisons with other null representatives.{ [DBNull]::Value } should not be equal to { [NullString]::Value }
```powershell
Expected $false, but got $true.
at <ScriptBlock>, /Users/vsts/agent/2.152.1/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 43
43:             $RHS.InvokeReturnAsIs() -eq $RHS.InvokeReturnAsIs() | Should -BeFalse
```

PoshChan commented:
@vexx32, your last commit had 5 failures in `PowerShell-CI-linux`
Null Representatives.Comparisons with $null.{ [DBNull]::Value } should be equivalent to $null (RHS: $null)
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 20
20:             $Value.InvokeReturnAsIs() -eq $null | Should -BeTrue
```
Null Representatives.Comparisons with $null.{ [NullString]::Value } should be equivalent to $null (RHS: $null)
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 20
20:             $Value.InvokeReturnAsIs() -eq $null | Should -BeTrue
```
Null Representatives.Comparisons with other null representatives.{ [AutomationNull]::Value } should not be equal to { [DBNull]::Value }
```powershell
Expected $false, but got $true.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 42
42:             $LHS.InvokeReturnAsIs() -eq $RHS.InvokeReturnAsIs() | Should -BeFalse
```
Null Representatives.Comparisons with other null representatives.{ [AutomationNull]::Value } should not be equal to { [NullString]::Value }
```powershell
Expected $false, but got $true.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 42
42:             $LHS.InvokeReturnAsIs() -eq $RHS.InvokeReturnAsIs() | Should -BeFalse
```
Null Representatives.Comparisons with other null representatives.{ [DBNull]::Value } should not be equal to { [NullString]::Value }
```powershell
Expected $false, but got $true.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 43
43:             $RHS.InvokeReturnAsIs() -eq $RHS.InvokeReturnAsIs() | Should -BeFalse
```

vexx32 commented:
All hail @SeeminglyScience's eyeballs tbh. I missed that both binder methods needed some extra logic there.

This should work, I think. Let me know if we need to make further changes @daxian-dbw but I think this ought to do it. Verified that the test cases added appear to pass locally so we should be set to go! 🤞

PoshChan commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -591,9 +591,7 @@ public static PSDataCollection<PSObject> GetPSDataCollection(object inputValue)
         /// <param name="second">Object to compare first to.</param>
```
@vexx32, your last commit had 4 failures in `PowerShell-CI-macos`
SemanticVersion api tests.Error handling.Missing parts: 'NullString': '$null'
```powershell
Expected an exception, with FullyQualifiedErrorId 'FormatException' to be thrown, but the FullyQualifiedErrorId was 'PSArgumentNullException'. from /Users/vsts/agent/2.153.2/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1:209 char:19
    +                 { [SemanticVersion]::Parse($version) } | Should -Thro ...
    +                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1: line 209
209:                 { [SemanticVersion]::Parse($version) } | Should -Throw -ErrorId "FormatException"
```
Null Representatives.Comparisons with other null representatives.{ [AutomationNull]::Value } should not be equal to { [DBNull]::Value }
```powershell
Expected $false, but got $true.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 42
42:             $LHS.InvokeReturnAsIs() -eq $RHS.InvokeReturnAsIs() | Should -BeFalse
```
Null Representatives.Comparisons with other null representatives.{ [AutomationNull]::Value } should not be equal to { [NullString]::Value }
```powershell
Expected $false, but got $true.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 42
42:             $LHS.InvokeReturnAsIs() -eq $RHS.InvokeReturnAsIs() | Should -BeFalse
```
Null Representatives.Comparisons with other null representatives.{ [DBNull]::Value } should not be equal to { [NullString]::Value }
```powershell
Expected $false, but got $true.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 43
43:             $RHS.InvokeReturnAsIs() -eq $RHS.InvokeReturnAsIs() | Should -BeFalse
```

PoshChan replied:
@vexx32, your last commit had 4 failures in `PowerShell-CI-linux`
SemanticVersion api tests.Error handling.Missing parts: 'NullString': '$null'
```powershell
Expected an exception, with FullyQualifiedErrorId 'FormatException' to be thrown, but the FullyQualifiedErrorId was 'PSArgumentNullException'. from /home/vsts/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1:209 char:19
    +                 { [SemanticVersion]::Parse($version) } | Should -Thro ...
    +                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1: line 209
209:                 { [SemanticVersion]::Parse($version) } | Should -Throw -ErrorId "FormatException"
```
Null Representatives.Comparisons with other null representatives.{ [AutomationNull]::Value } should not be equal to { [DBNull]::Value }
```powershell
Expected $false, but got $true.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 42
42:             $LHS.InvokeReturnAsIs() -eq $RHS.InvokeReturnAsIs() | Should -BeFalse
```
Null Representatives.Comparisons with other null representatives.{ [AutomationNull]::Value } should not be equal to { [NullString]::Value }
```powershell
Expected $false, but got $true.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 42
42:             $LHS.InvokeReturnAsIs() -eq $RHS.InvokeReturnAsIs() | Should -BeFalse
```
Null Representatives.Comparisons with other null representatives.{ [DBNull]::Value } should not be equal to { [NullString]::Value }
```powershell
Expected $false, but got $true.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 43
43:             $RHS.InvokeReturnAsIs() -eq $RHS.InvokeReturnAsIs() | Should -BeFalse
```

PoshChan replied:
@vexx32, your last commit had 2 failures in `PowerShell-CI-linux`
SemanticVersion api tests.Error handling.Missing parts: 'NullString': '$null'
```powershell
Expected an exception, with FullyQualifiedErrorId 'FormatException' to be thrown, but the FullyQualifiedErrorId was 'PSArgumentNullException'. from /home/vsts/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1:209 char:19
    +                 { [SemanticVersion]::Parse($version) } | Should -Thro ...
    +                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1: line 209
209:                 { [SemanticVersion]::Parse($version) } | Should -Throw -ErrorId "FormatException"
```
Null Representatives.Casting Behaviour.Error occurred in Context block
```powershell
Unable to find type [System.Management.Automation.AutomationNull].
At /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1:51 char:28
```

PoshChan replied:
@vexx32, your last commit had 2 failures in `PowerShell-CI-macos`
SemanticVersion api tests.Error handling.Missing parts: 'NullString': '$null'
```powershell
Expected an exception, with FullyQualifiedErrorId 'FormatException' to be thrown, but the FullyQualifiedErrorId was 'PSArgumentNullException'. from /Users/vsts/agent/2.153.2/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1:209 char:19
    +                 { [SemanticVersion]::Parse($version) } | Should -Thro ...
    +                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1: line 209
209:                 { [SemanticVersion]::Parse($version) } | Should -Throw -ErrorId "FormatException"
```
Null Representatives.Casting Behaviour.Error occurred in Context block
```powershell
Unable to find type [System.Management.Automation.AutomationNull].
At /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1:51 char:28
```

PoshChan replied:
@vexx32, your last commit had 5 failures in `PowerShell-CI-linux`
SemanticVersion api tests.Error handling.Missing parts: 'NullString': '$null'
```powershell
Expected an exception, with FullyQualifiedErrorId 'FormatException' to be thrown, but the FullyQualifiedErrorId was 'PSArgumentNullException'. from /home/vsts/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1:209 char:19
    +                 { [SemanticVersion]::Parse($version) } | Should -Thro ...
    +                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1: line 209
209:                 { [SemanticVersion]::Parse($version) } | Should -Throw -ErrorId "FormatException"
```
Null Representatives.Casting Behaviour.$null should be treated as $false by Where-Object
```powershell
Expected $null or empty, but got 100.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 64
64:             100 | Where-Object { $Value } | Should -BeNullOrEmpty
```
Null Representatives.Casting Behaviour.$null should be treated as $false by Where-Object
```powershell
Expected $null or empty, but got 100.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 64
64:             100 | Where-Object { $Value } | Should -BeNullOrEmpty
```
Null Representatives.Casting Behaviour.$null should cast to $false
```powershell
Expected $false, but got $true.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 58
58:             [bool]$Value | Should -BeFalse
```
Null Representatives.Casting Behaviour.$null should cast to $false
```powershell
Expected $false, but got $true.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 58
58:             [bool]$Value | Should -BeFalse
```

PoshChan replied:
@vexx32, your last commit had 1 failures in `PowerShell-CI-windows`
SemanticVersion api tests.Error handling.Missing parts: 'NullString': '$null'
```powershell
Expected an exception, with FullyQualifiedErrorId 'FormatException' to be thrown, but the FullyQualifiedErrorId was 'PSArgumentNullException'. from D:\a\1\s\test\powershell\engine\Basic\SemanticVersion.Tests.ps1:209 char:19
    +                 { [SemanticVersion]::Parse($version) } | Should -Thro ...
    +                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Basic\SemanticVersion.Tests.ps1: line 209
209:                 { [SemanticVersion]::Parse($version) } | Should -Throw -ErrorId "FormatException"
```

PoshChan replied:
@vexx32, your last commit had 1 failures in `PowerShell-CI-macos`
SemanticVersion api tests.Error handling.Missing parts: 'NullString': '$null'
```powershell
Expected an exception, with FullyQualifiedErrorId 'FormatException' to be thrown, but the FullyQualifiedErrorId was 'PSArgumentNullException'. from /Users/vsts/agent/2.153.2/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1:209 char:19
    +                 { [SemanticVersion]::Parse($version) } | Should -Thro ...
    +                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1: line 209
209:                 { [SemanticVersion]::Parse($version) } | Should -Throw -ErrorId "FormatException"
```

PoshChan replied:
@vexx32, your last commit had 1 failures in `PowerShell-CI-linux`
SemanticVersion api tests.Error handling.Missing parts: 'NullString': '$null'
```powershell
Expected an exception, with FullyQualifiedErrorId 'FormatException' to be thrown, but the FullyQualifiedErrorId was 'PSArgumentNullException'. from /home/vsts/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1:209 char:19
    +                 { [SemanticVersion]::Parse($version) } | Should -Thro ...
    +                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1: line 209
209:                 { [SemanticVersion]::Parse($version) } | Should -Throw -ErrorId "FormatException"
```

vexx32 commented:
Okay, I see.

So @daxian-dbw the comparisons such as `[AutomationNull]::Value -eq [DBNull]::Value` are returning `$true` because that's kind of how `[AutomationNull]` is intended to behave in comparisons, typically. It's treated like a true null would be.

Is it worth digging in here and trying to tease apart that logic to make these all mutually not equal each other, do you think? It would be inconsistent in terms of how comparisons to automation null typically behave.

I'm thinking the way it's currently functioning is pretty sensible, given the typical cases you'll come across with AutomationNull, having it behave just like comparing these null-like values directly to null behaves.

Would appreciate your thoughts here. 🙂

daxian-dbw commented:
> because that's kind of how `[AutomationNull]` is intended to behave in comparisons, typically. It's treated like a true null would be.

Good catch. Yes, `[AutomationNull]::Value` is supposed to behave just like `$null` with the only difference that `$null` can be passed through pipeline while `[AutomationNull]::Value` cannot.
Given that, I think it makes sense to have `[AutomationNull]::Value -eq [DBNull]::Value` returning `$true` (the same to `[AutomationNull]::Value -eq [NullString]::Value`).
@mklement0 Do you have any concerns?

vexx32 committed:
:sparkles: IsTrue() => false for all null-likes
:white_check_mark: Added test for cast and implicit convert to bool

vexx32 committed:
:white_check_mark: update comparison tests

PoshChan commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -591,9 +591,7 @@ public static PSDataCollection<PSObject> GetPSDataCollection(object inputValue)
         /// <param name="second">Object to compare first to.</param>
```
@vexx32, your last commit had 2 failures in `PowerShell-CI-windows`
SemanticVersion api tests.Error handling.Missing parts: 'NullString': '$null'
```powershell
Expected an exception, with FullyQualifiedErrorId 'FormatException' to be thrown, but the FullyQualifiedErrorId was 'PSArgumentNullException'. from D:\a\1\s\test\powershell\engine\Basic\SemanticVersion.Tests.ps1:209 char:19
    +                 { [SemanticVersion]::Parse($version) } | Should -Thro ...
    +                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Basic\SemanticVersion.Tests.ps1: line 209
209:                 { [SemanticVersion]::Parse($version) } | Should -Throw -ErrorId "FormatException"
```
Null Representatives.Casting Behaviour.Error occurred in Context block
```powershell
Unable to find type [System.Management.Automation.AutomationNull].
At D:\a\1\s\test\powershell\Language\Scripting\NullRepresentatives.Tests.ps1:51 char:28
```

vexx32 replied:
@daxian-dbw looking at the tests... It almost looks like this is working as intended?

Am I missing something there? Seems to me as though `ArgumentNullException` would be pretty much exactly what I'd expect a `NullString.Value` input to yield. 😕

vexx32 replied:
Ah, found the issue. The test's check for $null assumes that NullString.Value wouldn't pass the test -- but not it does. Fixed!

vexx32 committed:
:white_check_mark: Fix typo in nullrep tests

PoshChan commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -591,9 +591,7 @@ public static PSDataCollection<PSObject> GetPSDataCollection(object inputValue)
         /// <param name="second">Object to compare first to.</param>
```
@vexx32, your last commit had 5 failures in `PowerShell-CI-windows`
SemanticVersion api tests.Error handling.Missing parts: 'NullString': '$null'
```powershell
Expected an exception, with FullyQualifiedErrorId 'FormatException' to be thrown, but the FullyQualifiedErrorId was 'PSArgumentNullException'. from D:\a\1\s\test\powershell\engine\Basic\SemanticVersion.Tests.ps1:209 char:19
    +                 { [SemanticVersion]::Parse($version) } | Should -Thro ...
    +                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Basic\SemanticVersion.Tests.ps1: line 209
209:                 { [SemanticVersion]::Parse($version) } | Should -Throw -ErrorId "FormatException"
```
Null Representatives.Casting Behaviour.$null should be treated as $false by Where-Object
```powershell
Expected $null or empty, but got 100.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Scripting\NullRepresentatives.Tests.ps1: line 64
64:             100 | Where-Object { $Value } | Should -BeNullOrEmpty
```
Null Representatives.Casting Behaviour.$null should be treated as $false by Where-Object
```powershell
Expected $null or empty, but got 100.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Scripting\NullRepresentatives.Tests.ps1: line 64
64:             100 | Where-Object { $Value } | Should -BeNullOrEmpty
```
Null Representatives.Casting Behaviour.$null should cast to $false
```powershell
Expected $false, but got $true.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Scripting\NullRepresentatives.Tests.ps1: line 58
58:             [bool]$Value | Should -BeFalse
```
Null Representatives.Casting Behaviour.$null should cast to $false
```powershell
Expected $false, but got $true.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Scripting\NullRepresentatives.Tests.ps1: line 58
58:             [bool]$Value | Should -BeFalse
```

vexx32 replied:
🤔 I see I have some work yet to do. 😄

vexx32 committed:
:sparkles: Add IsNullLike() method
Checks the null-representatives as well as true nulls.

vexx32 committed:
:sparkles: fix conversions to boolean

vexx32 committed:
:recycle: refactor with helper method

SeeminglyScience commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -1065,12 +1064,20 @@ internal static bool IsTrue(IList objectArray)
+        /// </summary>
+        /// <param name="obj">The object to test.</param>
+        /// <returns>True if the object is null.</returns>
+        internal static bool IsNullLike(object obj) => obj == DBNull.Value || obj == NullString.Value || IsNull(obj);
```
What are the chances of this being public? The `LanguagePrimitives` methods are great for imitating PowerShell logic from C#.

If this was public we (external binary module writers) could:

1. Avoid doing all of those checks ourselves
1. Stop referencing the pseudo internal `AutomationNull`
1. Automatically handle new "null-like"'s when/if more are added

@daxian-dbw any objections?

vexx32 committed:
:tada: make IsNull and IsNullLike public

vexx32 committed:
:white_check_mark: Fix test null check
Test null check was assuming nullstring wouldn't pass the check.
Now it does, so we use the exposed IsNull() to resolve.

vexx32 committed:
:white_check_mark: Update tests
Output wasn't super readable with the way they were stringifying.
Wrap in {} and use InvokeReturnAsIs() to make the test result clearer.

vexx32 commented:
@daxian-dbw everything looks like it's all good.

@mklement, do you have any concerns about the comparison logic discussed here? (https://github.com/PowerShell/PowerShell/pull/9794#issuecomment-502417341 and https://github.com/PowerShell/PowerShell/pull/9794#issuecomment-502714647)

daxian-dbw commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -633,8 +629,7 @@ public static bool Equals(object first, object second, bool ignoreCase, IFormatP
 
-            var culture = formatProvider as CultureInfo;
-            if (culture == null)
+            if (!(formatProvider is CultureInfo culture))
```
Personally, I think the original `as` code is more readable here.

daxian-dbw commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -644,25 +639,38 @@ public static bool Equals(object first, object second, bool ignoreCase, IFormatP
+                    return true;
+                }
+
                 return false;
```
How about just `return IsNullLike(second);`

daxian-dbw commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -644,25 +639,38 @@ public static bool Equals(object first, object second, bool ignoreCase, IFormatP
+                    return true;
+                }
+
                 return false; // first is not null
```
Same here, maybe just `return IsNullLike(first);`

daxian-dbw commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -713,17 +721,17 @@ private static int CompareObjectToNull(object value, bool numberIsRightHandSide)
 
```
>  /// <param name="numberIsRightHandSide">True if the number to compare is on the right hand side **if** the comparison.</param>

Typo: should be **in**

daxian-dbw commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -777,23 +781,19 @@ public static int Compare(object first, object second, bool ignoreCase)
 
-            var culture = formatProvider as CultureInfo;
-            if (culture == null)
+            if (!(formatProvider is CultureInfo culture))
```
Same here. I think the `as` code is more readable in this case.

daxian-dbw commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -803,28 +803,34 @@ public static int Compare(object first, object second, bool ignoreCase, IFormatP
             {
-                string secondString = second as string;
-                if (secondString == null)
+                if (!(second is string secondString))
```
Same here.

daxian-dbw commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -713,17 +721,17 @@ private static int CompareObjectToNull(object value, bool numberIsRightHandSide)
-                case decimal de: return Math.Sign(de) < 0 ? -i : i;
-                default: return i;
-            }
+            return value switch
```
switch expression!!! Yay :)

vexx32 commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -3028,13 +3028,22 @@ private DynamicMetaObject BinaryBitwiseAnd(DynamicMetaObject target, DynamicMeta
             }
 
             var enumerable = PSEnumerableBinder.IsEnumerable(target);
             if (enumerable == null && arg.Value == null)
```
OK... it _looks_ like it should be done here somewhere?
https://github.com/vexx32/PowerShell/blob/ecccb0872fefb9e3c8e8b0949bf82a4ba0c96479/src/System.Management.Automation/engine/runtime/Binding/Binders.cs#L3377

I... don't know where to start there. 😅

vexx32 replied:
@daxian-dbw looks like we both spoke too soon. I have a lot of difficulty understanding all those code paths, but even without any editing yet, this works:

![image](https://user-images.githubusercontent.com/32407840/59977215-a52ae780-959c-11e9-86a5-df32e3c76cde.png)

So... I would _guess_ that somewhere in those paths we hook back into either LanguagePrimitives methods (which I've already handled) or back into the binder methods themselves as scalar comparisons.

I'll add some tests to make sure we don't regress. 😄

daxian-dbw replied:
Right :) I should have compile your branch and tried it myself.
https://github.com/PowerShell/PowerShell/blob/496741606736454c270bf97dc12a5a9eb73bca96/src/System.Management.Automation/engine/runtime/Binding/Binders.cs#L3360-L3362

Thanks for adding the collection comparison test!

vexx32 committed:
:recycle: pull null check into return statement

vexx32 committed:
:recycle: Address review comments

vexx32 committed:
:white_check_mark: Add collection comparison tests

mklement0 commented:
Thanks for asking, @daxian-dbw  and @vexx32, and sorry for the late reply.

The intricacies of the implementation are beyond me, but here's my end-user input:

Making `[AutomationNull]::Value` consistently interchangeably with `$null` makes sense, I suppose, but I do wish we could also make the following return `$true`, which currently isn't:

```
using namespace System.Management.Automation.Internal
[AutomationNull]::Value -is [AutomationNull]  # should be $true
```

On a related note - and that should probably be a separate issue - I wish that `Get-Member` acted differently:

```
Get-Member $noSuchVar  # should output $null - not fail, as it currently does
[System.Management.Automation.Internal.AutomationNull]::Value | Get-Member # should indicate that no pipeline input was received, not fail 
```

vexx32 commented:
Appreciate the input, @mklement0! Not sure about the latter, but seeing as the former already works with dbnull and nullstring it should probably be doable for automation null as well... Not sure exactly where that would need to be done, but I can take a look in a day or two :)

daxian-dbw commented:
@mklement0 Thanks for the input!

> `[AutomationNull]::Value -is [AutomationNull]  # should be $true`
> but seeing as the former already works with dbnull and nullstring it should probably be doable for automation null as well... Not sure exactly where that would need to be done, but I can take a look in a day or two :)

@vexx32 I suggest you to not work on supporting this in this PR. I'm not really sure we want to do that because it could cause confusion semantically -- `[AutomationNull]::Value` is treated as null in PowerShell (e.g. `[AutomationNull]::Value.GetType()` fails because it's not considered as an object), The current behavior is consistent and makes sense.

daxian-dbw reviewed (approved):
LGTM

mklement0 commented:
@daxian-dbw 

> [AutomationNull]::Value is treated as null in PowerShell (e.g. [AutomationNull]::Value.GetType()

Yes, but we're now also treating `[DbNull]::Value` and `[NullString]::Value` as `$null`, correct?

But they are distinct types with distinct behaviors and _sometimes_ you want to know what specific type you're dealing with.

And that currently only works (in a straightforward manner) with `[DbNull]::Value` and `[NullString]::Value` , with `-is <type>`.

By contrast, as shown above, it is far from trivial do distinguish between `$null` and `[AutomationNull]::Value`, yet that distinction arguably carries the most meaning in terms of behavior (sending something through the pipeline vs. not).

Do you not see value in detecting `[AutomationNull]::Value`, or do you envision a different solution?

vexx32 commented:
@daxian-dbw I do see your point there, and we can always add this later...

However. There already _is_ a disparity between `$null` and `[AutomationNull]` in that respect. For example:

```powershell
using namespace System.Management.Automation.Internal
$null -is [psobject] # returns $true
[AutomationNull]::Value -is [psobject] # returns $false
```

If we want AutomationNull to properly mimic $null as much as possible (minus the pipeline behaviour of course) it might possibly make _more_ sense to have it actually inherit PSObject or something along those lines. And for the few cases where we do actually need to distinguish $null from [AutomationNull] it would be helpful to have it also be properly recognised with `-is`. 😕

daxian-dbw commented:
> Yes, but we're now also treating `[DbNull]::Value` and `[NullString]::Value` as `$null`, correct?

No exactly, that's why the method added in `LanguagePrimitive` to check for them is called `IsNullLike` -- the change is only around when comparing `[DBNull]::Value` or `[NullString]::Value` with `$null` and `[AutomationNull]::Value`, and when casting `[DBNull]::Value` and `[NullString]::Value` to `bool`. `[System.DBNull]::Value.GetType()` still give you the type `DBNull`.

> Do you not see value in detecting [AutomationNull]::Value, or do you envision a different solution?

Personally, I don't see the value to differentiate `$null` and `[AutomationNull]::Value`. They should be the same, only with on difference that should be meaningful only to the engine -- `[AutomationNull]::Value` means not passing along the pipeline while `$null` does.

@vexx32 Thanks for showing that example (I didn't know it before :)). The comment you put in the code snippet are reversed though, `$null -is [psobject]` returns `false` and the other one returns `true`.
Yes, that's inconsistent, and how to fix that (if we want to) depends on how we want to treat `[AutomationNull]::Value` -- more of an implementation detail, or something that makes sense for users to depend on the be able to tell.

vexx32 commented:
Yeah, I noticed that went I went to check it myself; I've edited the comments. 😄 

Given there are circumstances where the difference between $null and [AutomationNull] _do_ matter to the user, I don't see any harm in permitting users to check for it explicitly with `-is`, really. If they really want to check it, there are already ways to do so, so I think we are better off permitting a more clear and explicit way to check for it, rather than forcing users to rely on this kind of hack:

```powershell
@($null).Count # outputs 1
@([AutomationNull]::Value).Count # outputs 0
```

(The above method is another way, of course, as well.)

daxian-dbw commented:
> `@([AutomationNull]::Value).Count # outputs 0`

Maybe we should fix this 😄, to make it more like a `$null`.

vexx32 commented:
Well, you say that, but... depending on the situation it can actually be kind of tricky. Because technically a pipeline that doesn't emit output ends up storing `[AutomationNull]::Value` iirc.

Quick example would be:

```powershell
$a = if ($False) { 2 }
@($a).Count # currently 0
```
Now, if we "fix" this, I think we may end up with a case where piping the `$a` value starts to invoke a pipeline and pass the `$null` value, or at least seem like it should.

The farther you go, the trickier it becomes. 😉

daxian-dbw commented:
It's definitely worth discussion and the behavior should depend on the real use cases where `[AutomationNull]::Value` has to be differentiated from `$null`.
I think you can open an issue to track the conversions around `[AutomationNull]::Value`. Once this PR is merged, the discussion here could be lost.

SeeminglyScience commented:
> @vexx32 Thanks for showing that example (I didn't know it before :)). The comment you put in the code snippet are reversed though, `$null -is [psobject]` returns `false` and the other one returns `true`.
> Yes, that's inconsistent, and how to fix that (if we want to) depends on how we want to treat `[AutomationNull]::Value` -- more of an implementation detail, or something that makes sense for users to depend on the be able to tell.

Ideally if this were to be fixed it would just be removed from assignment, e.g. `$obj = [AutomationNull]::Value` would populate `$obj` with true `null`.  The `-is` operator is one of the few things that never "lie" on occasion and I think it would be a detriment to change that.

You can also make `[AutomationNull]::Value -is [AutomationNull]` work by:

1. Removing the `static` keyword from the class decl
1. Make it inherit `PSObject`
1. Change the singleton instance to `new AutomationNull()`

That would allow `$autoNull -is [psobject]` to still work while enabling `$autoNull -is [AutomationNull]`.  Though I think there are a few places in the compiler where it uses a pattern like `if (expr.ExpressionType == typeof(PSObject) && expr.Value == AutomationNull.Value)` so that might need to change.

daxian-dbw commented:
@JamesWTruher Can you please quickly review this PR? Thanks!

vexx32 commented:
I'll write up a quick issue and link back here for the [automationnull] discussion & CC you guys so we can have a more focused discussion on that. 💖

JamesWTruher commented on test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1:
```diff
@@ -0,0 +1,83 @@
+
+Describe 'Null Representatives' -Tags 'CI' {
+
+    Context 'Comparisons with $null' {
```
although not specifically mentioned or explicitly tested, I assume that:

```powershell
if ( -not [dbnull]::Value ) {
   "nothing up my sleeves"
}
```
will work as expected and emit `nothing up my sleeves`

vexx32 replied:
Yep! `-not $var` forces a conversion to Boolean if the object isn't already a Boolean value.

I can add an explicit test if you feel it's warranted?

vexx32 replied:
Done! 🙂

vexx32 committed:
:white_check_mark: Add test case James mentioned

daxian-dbw commented:
@PoshChan Please remind me in 2 hours

PoshChan commented:
@daxian-dbw, this is the reminder you requested 2 hours ago

iSazonov commented:
@vexx32 Thanks for the great work!

daxian-dbw commented:
@LaurentDardenne Good catch. Will add this to breaking changes in the release log.

iSazonov commented:
@daxian-dbw  Should it be CL-BreakingChange?
